README.md
<img src="https://cdn.discordapp.com/avatars/804605929944645672/768065c543d1ed938858bf6f4ca81e15.png?size=4096" width="200" height="200" align="left" style="float: left; margin: 0 10px 0 0;" alt="Tanaka" />

# Tanaka

[![Discord](https://discord.com/api/guilds/830047984573480970/embed.png)](https://discord.gg/zGvtAnGhdP)
[![License](https://img.shields.io/github/license/1chiSensei/Tanaka)](https://github.com/1chiSensei/Tanaka/blob/main/LICENSE)
![Version](https://img.shields.io/github/release/1chiSensei/Tanaka.svg)
![Downloads](https://img.shields.io/github/downloads/1chiSensei/Tanaka/total.svg)
![Issues](https://img.shields.io/github/issues-raw/1chiSensei/Tanaka.svg)
![Pull Requests](https://img.shields.io/github/issues-pr/1chiSensei/Tanaka.svg)
![Contributors](https://img.shields.io/github/contributors/1chiSensei/Tanaka.svg)
![Last Commit](https://img.shields.io/github/last-commit/1chiSensei/Tanaka.svg)
![Stars](https://img.shields.io/github/stars/1chiSensei/Tanaka.svg?style=social&label=Stars&style=plastic)
![Website](https://img.shields.io/website-up-down-green-red/https/tanaka-bot.me.svg)
![](https://img.shields.io/badge/JavaScript-F7DF1E?logo=javascript&logoColor=black)

Tanaka is a Discord Bot coded in [JavaScript](https://www.javascript.com/) with [discord.js](https://discord.js.org) using the [Commando](https://github.com/discordjs/Commando) framework.
<br />
<br />

[![Deploy on Railway](https://railway.app/button.svg)](https://railway.app/new/template?template=https%3A%2F%2Fgithub.com%2F1chiSensei%2FTanaka&plugins=redis%2Cmongodb&envs=CLIENT_ID%2CCLIENT_SECRET%2CCOMMAND_PREFIX%2CDISCORD_BOTS_GG_KEY%2CDISCORD_TOKEN%2CGOOGLE_KEY%2CIBL_KEY%2COCR_KEY%2COWNER_ID%2CREDDIT_CLIENT_ID%2CREDDIT_CLIENT_SECRET%2CREDDIT_PASSWORD%2CSTACKOVERFLOW_KEY%2CTEST_WEBHOOK_ID%2CTEST_WEBHOOK_TOKEN&CLIENT_IDDesc=The+application+client+id.&CLIENT_SECRETDesc=The+bot%27s+client+secret.&COMMAND_PREFIXDesc=The+prefix+of+the+bot.&DISCORD_BOTS_GG_KEYDesc=The+api+key+for+discord.bots.gg.&DISCORD_TOKENDesc=The+bot%27s+token.&GOOGLE_KEYDesc=Your+google+api+key.&IBL_KEYDesc=The+api+key+for+infinitybotlist.com.&OCR_KEYDesc=The+api+key+for+https%3A%2F%2Focr.space.&OWNER_IDDesc=Your+own+discord+id.&REDDIT_CLIENT_IDDesc=Your+reddit+client+id.&REDDIT_CLIENT_SECRETDesc=Your+reddit+client+secret.&REDDIT_PASSWORDDesc=Your+reddit+password.&STACKOVERFLOW_KEYDesc=Your+stack+overflow+api+key.&TEST_WEBHOOK_IDDesc=The+test+webhook+id.&TEST_WEBHOOK_TOKENDesc=The+test+webhook+token.&COMMAND_PREFIXDefault=t%21&OWNER_IDDefault=566155739652030465)

## Copyright

- ©2021 [Yoshida Tomio](https://github.com/1chiSensei)

## Invite

- [Support Server](https://discord.gg/zGvtAnGhdP)
- [Bot Invite](https://peico.xyz/T4DQP)

## Permissions

- **Manage Webhooks** is required for the `anime-updates` command.
- **Create Instant Invite** is required for the `serverinfo` command.
- **Change Nickname** is optional. (But recommended.)
- **Read Messages** is required for every command.
- **Send Messages** is required for every command.
- **Embed Links** is required for every command that sends an embed.
- **Attach Files** is required for every command that attaches a file.
- **Add Reactions** is required for every command that sends a reaction.
- **Use External Emojis** is required for every command.
- **Read Message History** is required for every command.
- **Send TTS Message** is required for every command that sends a TTS message.
- **View Channel** is required for every command.
- **Connect** is required for music commands.
- **Speak** is required for music commands.
- **Use Voice Activity** is required for music commands.
- **Priority Speaker** is required for music commands.

## Installing

You can't. You just can't. There are alot of keys needed for this bot and you just can't get them.

## Commands

Total: 50

### Utility:

- **prefix:** Shows or sets the command prefix.
- **eval:** Executes arbitrary JavaScript code. (Owner-Only)
- **exec:** Executes a command line application. (Owner-Only)
- **ping:** Shows you the latency of the bot and the database.
- **restart:** Restarts the bot. (Owner-Only)

### Random Response:

- **advice:** Responds with a random advice.
- **bird:** Responds with a random bird image.
- **bunny:** Responds with a random bunny image.
- **cat:** Responds with a random cat image.
- **dog:** Responds with a random dog image.
- **duck:** Responds with a random duck image.
- **fact:** Responds with a random fact.
- **fml:** Responds with a random FML quote.
- **fox:** Responds with a random fox image.
- **koala:** Responds with a random panda image.
- **lizard:** Responds with a random lizard image.
- **owl:** Responds with a random owl image.
- **panda:** Responds with a random panda image.
- **shiba:** Responds with a random shiba inu image.

### Information:

- **help:** Displays a list of available commands, or detailed information for a specific command.
- **info:** Responds with the general information of the bot.
- **serverinfo:** Responds with the server's general information.
- **support:** Sends the support server of the bot.
- **userinfo:** Responds with the user's general information.

### Search:

- **anime:** Searches MyAnimeList for the anime that you give.
- **mdn:** Searches MDN for your query.
- **screenshot:** Screenshots a website.
- **stack-overflow:** Searches Stack Overflow for your query.

### Reminder:

- **delete-reminder:** Deletes your reminder.
- **remind:** Sets a reminder.

### Anime Updates:

- **anime-updates:** Sets the anime updates channel.
- **delete-update-channel:** Deletes the anime updates channel.

### Code Bins:

- **hastebin:** Uploades your code to https://hastebin.com.
- **nopaste:** Uploades your code to https://nopaste.ml.
- **sourcebin:** Uploades your code to https://sourceb.in.
- **starbin:** Uploades your code to https://starb.in.

### Image Manipulation:

- **glitch:** Applies the glitch effect to an image.
- **triggered:** Applies the triggered effect to an image.

### Music:

- **pause:** Pauses the music queue.
- **play:** Adds music to the music queue.
- **queue:** Shows the music queue.
- **resume:** Resumes the music queue.
- **skip:** Skips the music queue.
- **stop:** Stops the music queue.

### Commands:

- **groups:** Lists all command groups.
- **enable:** Enables a command or command group.
- **disable:** Disables a command or command group.
- **reload:** Reloads a command or command group. (Owner-Only)
- **load:** Loads a new command. (Owner-Only)
- **unload:** Unloads a command. (Owner-Only)

## Star Chart

![star chart](https://starchart.cc/1chiSensei/Tanaka.svg)


CHANGELOG.md
## [2.97.1](https://github.com/1chiSensei/Tanaka/compare/2.97.0...2.97.1) (2021-06-09)

### Bug Fixes

- **ci:** fix ([c98fdb3](https://github.com/1chiSensei/Tanaka/commit/c98fdb30a983a06b2221adf28462c00930d90df0))
- **global:** fix bot not responding ([b2941ed](https://github.com/1chiSensei/Tanaka/commit/b2941edcc0cd0b4c0817cb689e5939264c64a077))
- **global:** fix error ([58ff959](https://github.com/1chiSensei/Tanaka/commit/58ff95920b124c4598d97846c542b1289129fbe6))

# [2.97.0](https://github.com/1chiSensei/Tanaka/compare/2.96.0...2.97.0) (2021-06-04)

### Features

- **docs:** add example env ([b3add92](https://github.com/1chiSensei/Tanaka/commit/b3add9210b06f0e90c6b40d72fccdb31d6e74f76))

# [2.96.0](https://github.com/1chiSensei/Tanaka/compare/2.95.0...2.96.0) (2021-06-04)

### Features

- **sentry:** add sentry ([50057e7](https://github.com/1chiSensei/Tanaka/commit/50057e75ec7f48b15de28e059a0dfe20a15f1115))

# [2.95.0](https://github.com/1chiSensei/Tanaka/compare/2.94.1...2.95.0) (2021-06-04)

### Features

- **events:** send info message when a bot joins a guild ([1306497](https://github.com/1chiSensei/Tanaka/commit/1306497619eb5d843c4138dda9797ccacfeaa962))

## [2.94.1](https://github.com/1chiSensei/Tanaka/compare/2.94.0...2.94.1) (2021-06-04)

### Bug Fixes

- **web:** fix command run count ([c46f136](https://github.com/1chiSensei/Tanaka/commit/c46f1362f8487a7a98407a92d1eb6bb474b50906))

# [2.94.0](https://github.com/1chiSensei/Tanaka/compare/2.93.0...2.94.0) (2021-06-04)

### Features

- **commands:** add a base command class ([83e3a08](https://github.com/1chiSensei/Tanaka/commit/83e3a08664320f375a8364505f88accd3f0f8a98))

# [2.93.0](https://github.com/1chiSensei/Tanaka/compare/2.92.0...2.93.0) (2021-06-04)

### Features

- **web:** add normalize.css ([421dbf9](https://github.com/1chiSensei/Tanaka/commit/421dbf9b3b9534044bbb82f7c04e19b2a4e67c6d))

# [2.92.0](https://github.com/1chiSensei/Tanaka/compare/2.90.0...2.92.0) (2021-06-04)

### Bug Fixes

- **ci:** fix ([5c2d4f4](https://github.com/1chiSensei/Tanaka/commit/5c2d4f4afc139d39018053dc04e3be6a91357185))
- **ci:** fix ([f295b0e](https://github.com/1chiSensei/Tanaka/commit/f295b0ee6352f4e706d326b0bc0c7f7b0a4ca0dd))

### Features

- **ci:** fix ([aa18e97](https://github.com/1chiSensei/Tanaka/commit/aa18e97a8df186fe0223000c45fe467ac3ef847e))
- **web:** add break lines ([e144f19](https://github.com/1chiSensei/Tanaka/commit/e144f1925d2742e8d0179111f21176d1bdeb1f71))
- **web:** add more info ([9139ece](https://github.com/1chiSensei/Tanaka/commit/9139ecef20e1396c3c177179a482cdc4a85d061d))

# [2.91.0](https://github.com/1chiSensei/Tanaka/compare/2.90.0...2.91.0) (2021-06-04)

### Bug Fixes

- **ci:** fix ([5c2d4f4](https://github.com/1chiSensei/Tanaka/commit/5c2d4f4afc139d39018053dc04e3be6a91357185))
- **ci:** fix ([f295b0e](https://github.com/1chiSensei/Tanaka/commit/f295b0ee6352f4e706d326b0bc0c7f7b0a4ca0dd))

### Features

- **ci:** fix ([aa18e97](https://github.com/1chiSensei/Tanaka/commit/aa18e97a8df186fe0223000c45fe467ac3ef847e))
- **web:** add more info ([9139ece](https://github.com/1chiSensei/Tanaka/commit/9139ecef20e1396c3c177179a482cdc4a85d061d))

# [2.90.0](https://github.com/1chiSensei/Tanaka/compare/2.89.2...2.90.0) (2021-06-04)

### Features

- **commands:** add spacebin command ([b82bcbc](https://github.com/1chiSensei/Tanaka/commit/b82bcbcfb5afb0186b7f751c2d7bbb178a8f9942))

## [2.89.2](https://github.com/1chiSensei/Tanaka/compare/2.89.1...2.89.2) (2021-06-04)

### Bug Fixes

- **commands:** add missing emoji ([1ad6ce9](https://github.com/1chiSensei/Tanaka/commit/1ad6ce9519d575fc3937b1abba3eac8b435c8db8))
- **yarn:** clean cache ([5261615](https://github.com/1chiSensei/Tanaka/commit/526161528de1b4ad2fe9b7c29a982184f9ad8895))

## [2.89.1](https://github.com/1chiSensei/Tanaka/compare/2.89.0...2.89.1) (2021-06-04)

### Bug Fixes

- **scripts:** fix ([df906aa](https://github.com/1chiSensei/Tanaka/commit/df906aa519b201250f01097ea7257f771c1364ee))

# [2.89.0](https://github.com/1chiSensei/Tanaka/compare/2.88.0...2.89.0) (2021-06-03)

### Features

- **scripts:** start ([ff1b9b7](https://github.com/1chiSensei/Tanaka/commit/ff1b9b7fe0550286e63b0ca3d86d47ef93bc32cd))

# [2.88.0](https://github.com/1chiSensei/Tanaka/compare/2.87.0...2.88.0) (2021-06-03)

### Features

- **commands:** use emojis ([4e041a2](https://github.com/1chiSensei/Tanaka/commit/4e041a2561f6feace808e9b45f3919395702c273))

# [2.87.0](https://github.com/1chiSensei/Tanaka/compare/2.86.1...2.87.0) (2021-06-03)

### Features

- **commands:** use embeds ([9aebe28](https://github.com/1chiSensei/Tanaka/commit/9aebe28d9bd25868ac7e2078e6a55ba452056885))

## [2.86.1](https://github.com/1chiSensei/Tanaka/compare/2.86.0...2.86.1) (2021-06-03)

### Bug Fixes

- **commands:** fix string ([f7d1bdf](https://github.com/1chiSensei/Tanaka/commit/f7d1bdf5ff0d7152dd55b4945a7b863cad6afa15))

# [2.86.0](https://github.com/1chiSensei/Tanaka/compare/2.85.2...2.86.0) (2021-06-03)

### Features

- **commands:** make eval show the type ([9d10766](https://github.com/1chiSensei/Tanaka/commit/9d1076650b783cb3387113db0e94cc981bf9d928))

## [2.85.2](https://github.com/1chiSensei/Tanaka/compare/2.85.1...2.85.2) (2021-06-03)

### Bug Fixes

- **proxy:** remove letsencrypt field ([dd8e830](https://github.com/1chiSensei/Tanaka/commit/dd8e830226abac3041cd732143191f4507f501cc))

## [2.85.1](https://github.com/1chiSensei/Tanaka/compare/2.85.0...2.85.1) (2021-06-03)

### Bug Fixes

- **proxy:** fix port error ([9effff9](https://github.com/1chiSensei/Tanaka/commit/9effff9d2d4ed1480f2aa367c4da544b20663480))

# [2.85.0](https://github.com/1chiSensei/Tanaka/compare/2.84.1...2.85.0) (2021-06-03)

### Features

- **web:** use redbird for proxy ([53cb01c](https://github.com/1chiSensei/Tanaka/commit/53cb01c65e481998038abb05af77d58bc016c19d))

## [2.84.1](https://github.com/1chiSensei/Tanaka/compare/2.84.0...2.84.1) (2021-06-03)

### Bug Fixes


src/Web/index.js
const { MessageEmbed, WebhookClient } = require('discord.js');
const manifest = require('./manifest.json');
const { Client } = require('@2pg/oauth');
const express = require('express');
const cookies = require('cookies');
const helmet = require('helmet');
const morgan = require('morgan');
const path = require('path');
const cors = require('cors');

const app = express();
const client = new Client({
	id: process.env.CLIENT_ID,
	secret: process.env.CLIENT_SECRET,
	redirectURI: 'https://tanaka-bot.me/auth/callback',
	scopes: ['identify'],
});
const port = 3001;
const hook = new WebhookClient(process.env.VOTE_ID, process.env.VOTE_TOKEN);

// eslint-disable-next-line valid-jsdoc
/**
 *  @param {import('../Structures/Client')} c The client
 */
module.exports = (c) => {
	app.use(cors());
	app.use(helmet({ contentSecurityPolicy: false, frameguard: false }));
	app.use(express.json());
	app.use(express.urlencoded({ extended: true }));
	app.use(express.text());
	app.use(cookies.express(['I', 'store', 'login', 'sessions.']));
	app.use(morgan('combined'));
	app.use((_, res, next) => {
		res.header('Service-Worker-Allowed', '/');
		next();
	});
	app.use('/static', express.static(path.join(__dirname, 'static')));

	app.set('view engine', 'ejs');
	app.set('trust proxy', true);
	app.set('json spaces', 8);
	app.set('views', path.join(__dirname, 'views'));

	app.get('/', async (req, res) => {
		const key = req.cookies.get('discordToken');

		const userCount = (await c.userCount()).toLocaleString();
		const guildCount = (await c.guildCount()).toLocaleString();
		const channelCount = (await c.channelCount()).toLocaleString();
		const usesCount = (await c.db.get('commandRunCount')).toLocaleString();
		const commandCount = c.registry.commands.size.toLocaleString();
		const shardCount = c.shard ? c.shard.shardCount.toLocaleString() : '1';
		const ping = Math.round(c.ws.ping);

		const data = key ? await client.getUser(key) : null;

		res.render('index', {
			data,
			userCount,
			guildCount,
			channelCount,
			commandCount,
			shardCount,
			ping,
			usesCount,
		});
	});

	app.get('/invite', (_, res) =>
		res.redirect(
			302,
			'https://discord.com/api/oauth2/authorize?client_id=804605929944645672&permissions=379968&redirect_uri=https%3A%2F%2Ftanaka-bot.me%2Fauth%2Fcallback&response_type=code&scope=identify%20bot',
		),
	);

	app.get('/discord', (_, res) => res.redirect(302, 'https://discord.gg/zGvtAnGhdP'));

	app.get('/github', (_, res) => res.redirect(302, 'https://github.com/1chiSensei/Tanaka'));

	app.get('/commands', async (req, res) => {
		const key = req.cookies.get('discordToken');

		res.render('commands', {
			data: key ? await client.getUser(key) : null,
			commands: c.generateCommandList(),
		});
	});

	app.get('/legal', async (req, res) => {
		const key = req.cookies.get('discordToken');

		res.render('legal', { data: key ? await client.getUser(key) : null });
	});

	app.get('/vote', (_, res) => res.render('vote'));

	app.get('/profile', async (req, res) => {
		const key = req.cookies.get('discordToken');

		res.render('profile', {
			data: key ? await client.getUser(key) : null,
			moment: require('moment'),
		});
	});

	app.get('/auth/login', (_, res) => res.redirect(302, client.authCodeLink.url));

	app.get('/auth/callback', async (req, res) => {
		if (req.cookies.get('discordToken')) return res.redirect(302, '/');

		const code = req.query.code.toString();
		const key = await client.getAccess(code);

		res.cookie('discordToken', key).redirect('/');
	});

	app.get('/auth/logout', (_, res) => {
		res.clearCookie('discordToken', undefined);

		res.redirect(302, '/');
	});

	app.post('/vote/ibl', async (req, res) => {
		if (req.header('Authorization') !== process.env.VOTE_AUTH) {
			if (req.cookies.get('discordToken')) return res.sendStatus(403);
			return res.sendStatus(401);
		}

		const { userID } = req.body;
		const user = await c.users.fetch(userID);
		const embed = new MessageEmbed()
			.setTitle('New Vote!')
			.setURL('https://infinitybotlist.com/bots/804605929944645672')
			.setThumbnail(user.displayAvatarURL({ dynamic: true, size: 4096 }))
			.setDescription(`Thank you, ${user.tag}, for voting for Tanaka in Infinity Bot List!`)
			.setColor('RANDOM')
			.setFooter('infinitybotlist.com')
			.setTimestamp();

		await hook.send(embed);

		res.sendStatus(201);
	});

	app.post('/vote/voidbots', async (req, res) => {
		if (req.header('Authorization') !== process.env.VOTE_AUTH) {
			if (req.cookies.get('discordToken')) return res.sendStatus(403);
			return res.sendStatus(401);
		}

		const { user: userID } = req.body;
		const user = await c.users.fetch(userID);
		const embed = new MessageEmbed()
			.setTitle('New Vote!')
			.setURL('https://voidbots.net/bot/804605929944645672')
			.setThumbnail(user.displayAvatarURL({ dynamic: true, size: 4096 }))
			.setDescription(`Thank you, ${user.taf}, for voting for Tanaka in Void Bots!`)
			.setColor('RANDOM')
			.setFooter('voidbots.net')
			.setTimestamp();

		await hook.send(embed);

		res.sendStatus(201);
	});

	app.get('/manifest.json', (_, res) => res.status(200).json(manifest));

	app.get('/serviceWorker.js', (_, res) =>
		res.status(200).sendFile(path.join(__dirname, 'static', 'serviceWorker.js')),
	);

	app.listen(port, () => c.logger.info(`Listening on localhost:${port}`));
};
